[Gu Haiwei] ip

WilliamHaiweiGu opened the pull request:
> You just got REKT by the upcoming star,
# Meggy
Command line based task list manager. Name based on a character created by YouTuber [SMG4](https://www.youtube.com/@SMG4).
Implemented commands:
1. bye
1. list
1. mark
1. unmark
1. todo
1. deadline
1. event
1. delete
1. find

⚠️Note that:
* Delete operation is currently **O(n)**
* Text UI so far

Most commands, messages, and keyword strings are customizable and stored in the `Resource` class:
```java
public final static String ERR_BASE = " OOPS!!! ";
public final static String ERR_NO_ARGS = "Description of this command cannot be empty.\n";
public final static String ERR_FILE_WRITE = "Can't save data to file. ";
public final static String ERR_IO = "An IO error occurred.\n";
public final static String ERR_NO_FILE_ACCESS = "File access denied by system.\n";
public final static String TASK_STRING_INDENT = "    ";
public final static String CMD_EXIT = "bye";
public final static String CMD_LIST = "list";
public final static String CMD_MARK = "mark";
public final static String CMD_UNMK = "unmark";
public final static String CMD_TODO = "todo";
public final static String CMD_DDL = "deadline";
public final static String CMD_EVENT = "event";
public final static String CMD_DEL = "delete";
public final static String CMD_FIND = "find";
public final static String KW_DUE = "by";
public final static String KW_STT = "from";
public final static String KW_END = "to";
public final static String NOTIF_ADD = " Got it. Added this task:\n";
public final static String NOTIF_LIST = " Here are the tasks in your list:\n";
public final static String NOTIF_MARK = " Booyah! Marked this task as done:\n";
public final static String NOTIF_UNMK = " OK. Marked this task as not done:\n";
public final static String NOTIF_USAGE = " Usage: ";
public final static String NOTIF_DEL = " OK. Removed this task:\n";
public final static String NOTIF_FIND = " Here are the matching tasks in your list:\n";
```

Features to implement next:
- [ ] JavaFX GUI
- [ ] Use balanced binary tree to make deletion O(log(n))

Commit:
Add Gradle support

WilliamHaiweiGu committed:
Echo and bye implemented

WilliamHaiweiGu committed:
add and list implemented

WilliamHaiweiGu committed:
Marking and Unmarking implemented

WilliamHaiweiGu committed:
Todo Task implemented

WilliamHaiweiGu committed:
DDL Task implemented

WilliamHaiweiGu committed:
Event Task implemented

WilliamHaiweiGu committed:
Text UI tested with io streams

WilliamHaiweiGu committed:
Index parsing error handled

WilliamHaiweiGu committed:
Parsing errors handled with MeggyException classes

WilliamHaiweiGu committed:
O(n) task deletion implemented

WilliamHaiweiGu committed:
Added documentation

WilliamHaiweiGu committed:
Text UI Test 2

WilliamHaiweiGu committed:
Implement task list data file (MeggyData.txt) saving and loading.

WilliamHaiweiGu committed:
Implement MeggyTime - formatted or unformatted time

WilliamHaiweiGu committed:
Merge branch 'branch-Level-8'

WilliamHaiweiGu committed:
Merge branch Level-8. Add encode to MeggyTime

WilliamHaiweiGu committed:
Merge remote-tracking branch 'origin/add-gradle-support'

WilliamHaiweiGu committed:
Add Gradle Support. Can build, run, and clean using gradle panel

WilliamHaiweiGu committed:
Extract UI class

WilliamHaiweiGu committed:
Add more UI documentation

WilliamHaiweiGu committed:
Extract Storage class

WilliamHaiweiGu committed:
Extract Parser class

WilliamHaiweiGu committed:
Extract TaskList class. Move bounds check out of index parsing

WilliamHaiweiGu committed:
Organize the classes into packages

WilliamHaiweiGu committed:
Use Gradle for testing. Add JUnit test. Add equals method to all UserTask classes.

WilliamHaiweiGu committed:
Code style comply with standard

WilliamHaiweiGu committed:
Update gradle.build mainClassName so that JAR export works correctly

WilliamHaiweiGu committed:
Improve documentation

WilliamHaiweiGu committed:
Improve coding standard

WilliamHaiweiGu committed:
Implement 'find' command. Fix bug of blank arguments

WilliamHaiweiGu committed:
Merge branch 'branch-A-JavaDoc'

WilliamHaiweiGu committed:
Merge branch 'branch-A-CodingStandard'

WilliamHaiweiGu committed:
Merge branch 'branch-Level-9'

WilliamHaiweiGu committed:
Fix documentation and coding standard

tyx021 commented on src/main/java/Meggy/Resource.java:
```diff
@@ -0,0 +1,126 @@
+     * @return Message about empty input or unknown command or empty input.
+     */
+    public static String errUnknownCmd(String cmd) {
+        return "".equals(cmd) ? "Umm that's all white space? Say something! Speak to me!\n" : "Don't know what \"" +
+                cmd + "\" means.\n";
```
Maybe you can try to have a line break before '+' operator? 🤔 

```suggestion
        return "".equals(cmd) ? "Umm that's all white space? Say something! Speak to me!\n" : "Don't know what \""
                + cmd + "\" means.\n";
```

tyx021 commented on src/main/java/Meggy/Meggy.java:
```diff
@@ -0,0 +1,186 @@
+package Meggy;
+
+import Meggy.Exception.Function;
+import Meggy.Exception.MeggyException;
+import Meggy.Exception.MeggyNoArgException;
+import Meggy.Task.UserTask;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+import java.util.Scanner;
```
Nice! You have followed the convention of importing statements!

tyx021 commented on src/main/java/Meggy/Resource.java:
```diff
@@ -0,0 +1,126 @@
+     * @param arg Non-null. String in question.
+     * @return Error message.
+     */
+    public static String errNFE(String arg) {
```
Perhaps it will be better to name this method as errNfe()?
```suggestion
    public static String errNfe(String arg) {
```

tyx021 commented on src/main/java/MeggyTest.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private static MeggyTime randMeggyTime() {
+        return MeggyTime.of(RAND.nextDouble() < 0.5 ? randString()
+                : LocalDateTime.ofEpochSecond(RAND.nextInt(), 0, ZoneOffset.UTC).format(MeggyTime.ENCODE_FORMAT));
```
Maybe try to write ternary expressions in this form?
```suggestion
        return MeggyTime.of(RAND.nextDouble() < 0.5 
                ? randString()
                : LocalDateTime.ofEpochSecond(RAND.nextInt(), 0, ZoneOffset.UTC).format(MeggyTime.ENCODE_FORMAT));
```

tyx021 commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,175 @@
+ 4.[D][ ] return book (by: [Sunday])
+ 5.[D][ ] return book (by: 2019-12-02 18:00)
+------------------------------------------------------------
+
+>------------------------------------------------------------
+ OK gotta go play more Turf Wars. Have a nice day!
+------------------------------------------------------------
+
```
I like your test cases! It tests all the functions and error handling! 😄

tyx021 commented on src/main/java/Meggy/Util.java:
```diff
@@ -0,0 +1,40 @@
+
+    /** Puts {@link Character}s in square brackets. */
+    public static String parenthesize(char c) {
+        return "[" + c + ']';
```
Is there any reason why you used two different type of quotes here? 🤔

tyx021 reviewed (commented):
Overall, I like how you handle the flow logic of the bot elegantly! 👍 👍 
I believe the detailed error message will be very helpful for the user 🙏 
And your bot has very interesting personality as well! 😄 Good job!

WilliamHaiweiGu committed:
Use stylechecker. Improve coding standard

WilliamHaiweiGu committed:
Fully integrate GUI

WilliamHaiweiGu committed:
Merge branch 'branch-Level-10'

WilliamHaiweiGu committed:
Use java asssert statements to check for preconditions.

Many methods requires non-null parameters. Before, they were only noted
on documentation. Now, using assert statements to check nullity of all
non-null parameters.

Found bug caused by a highly unusually user input combination. Recording
in bug trcker for now.

WilliamHaiweiGu committed:
Comply with coding standard.

Removed magic numbers with well-named constants.

Added comment to all empty catch blocks.

Wrapped lines longer than 120 chars.

WilliamHaiweiGu committed:
Merge pull request #1 from WilliamHaiweiGu/branch-A-Assertions

Use java asssert statements to check for preconditions.

WilliamHaiweiGu committed:
Merge branch 'master' into branch-A-CodeQuality

WilliamHaiweiGu committed:
Merge pull request #2 from WilliamHaiweiGu/branch-A-CodeQuality

Comply with coding standard.

WilliamHaiweiGu committed:
Implement duplication detection. Duplicated tasks won't be added.

WilliamHaiweiGu committed:
Fix GUI spacing bug

WilliamHaiweiGu committed:
Install Splatoon fonts

WilliamHaiweiGu committed:
Handle null font error

WilliamHaiweiGu committed:
Make GUI resizable

WilliamHaiweiGu committed:
Tweak padding, fonts

WilliamHaiweiGu committed:
Make storage module keep task completion status

WilliamHaiweiGu committed:
Add notification channel. Fix bug of nitification message replacing normal response

WilliamHaiweiGu committed:
Fix deadline task bug of adding space to desciption. No longer buggy if desc ends with keyword

WilliamHaiweiGu committed:
Remove command regeneration. Store the creation command and look it up in every recreateCmd() call

WilliamHaiweiGu committed:
Add storage testing

WilliamHaiweiGu committed:
Improve code quality

WilliamHaiweiGu committed:
Add icon and title. Remove code duplication. Feature finailzed

WilliamHaiweiGu committed:
Add User Guide. Minor code improvement

WilliamHaiweiGu committed:
Rename UI screenshot file

WilliamHaiweiGu committed:
Fix wrong testcases. Pass file into Meggy constructor

WilliamHaiweiGu committed:
Improve user guide & code quality

WilliamHaiweiGu committed:
Add font info. Pass 100 CPU-hour of testing
